src/GraphQl/Resolver/Factory/CollectionResolverFactory.php

```diff
@@ -84,17 +91,38 @@ public function __invoke(string $resourceClass = null, string $rootClass = null,
             if (isset($rootClass, $source[$rootProperty = $info->fieldName], $source[ItemNormalizer::ITEM_KEY])) {
                 $rootResolvedFields = $this->identifiersExtractor->getIdentifiersFromItem(unserialize($source[ItemNormalizer::ITEM_KEY]));
                 $subresource = $this->getSubresource($rootClass, $rootResolvedFields, array_keys($rootResolvedFields), $rootProperty, $resourceClass, true, $dataProviderContext);
+
+                if (null !== $this->dispatcher) {
+                    $this->dispatcher->dispatch(PreReadEvent::NAME, new PreReadEvent($subresource));
+                }
+
                 $collection = $subresource ?? [];
             } else {
+                if (null !== $this->dispatcher) {
+                    $this->dispatcher->dispatch(PreReadEvent::NAME, new PreReadEvent([]));
+                }
+
                 $collection = $this->collectionDataProvider->getCollection($resourceClass, null, $dataProviderContext);
             }
 
+            if (null !== $this->dispatcher) {
+                $this->dispatcher->dispatch(PostReadEvent::NAME, new PostReadEvent($collection));
+            }
+
             $this->canAccess($this->resourceAccessChecker, $resourceMetadata, $resourceClass, $info, $collection, $operationName ?? 'query');
 
             if (!$this->paginationEnabled) {
                 $data = [];
                 foreach ($collection as $index => $object) {
+                    if (null !== $this->dispatcher) {
+                        $this->dispatcher->dispatch(PreSerializeEvent::NAME, new PreSerializeEvent($collection));
+                    }
+
                     $data[$index] = $this->normalizer->normalize($object, ItemNormalizer::FORMAT, $dataProviderContext);
+
+                    if (null !== $this->dispatcher) {
+                        $this->dispatcher->dispatch(PostSerializeEvent::NAME, new PostSerializeEvent($data[$index]));
+                    }
```

Comment:
dispatching every item in the collection looks heavy, can't we jus dispatch the collection?